api/app/utils/fba_calculator.py

```diff
@@ -191,108 +186,133 @@ def get_approx_flame_length(head_fire_intensity: float):
+        # while get_morning_diurnal_ffmc(clock_time, yesterdays_ffmc, relative_humidity) >= critical_ffmc:
+        #     clock_time -= 0.5
+        #     if clock_time == -0.5:
+        #         break
+        # # add back the half hour that caused FFMC to drop below critical_ffmc (or that
+        # #   pushed time below 0.0)
+        # clock_time += 0.5
+        # logger.info('%s', clock_time)
+        # return clock_time
+
+        # NOTE: For now, simply returning 13.0 hours until we're able to calculate diurnal FFMC for
+        # morning hours.
+        return 13.0
+
+    logger.info('Solar noon FFMC %s < critical FFMC %s', solar_noon_ffmc, critical_ffmc)
+    # go forward in time in increments of 1 hour
+    clock_time = 13 + 1.0  # start from solar noon + 1 hours
+    while get_afternoon_overnight_diurnal_ffmc(clock_time, solar_noon_ffmc) < critical_ffmc:
+        clock_time += 1.0
+        if clock_time == 24.0:
+            return None
+    return clock_time
+
+
+def get_critical_hours_end(critical_ffmc: float, solar_noon_ffmc: float, critical_hour_start: float):
+    """ Returns the hour of day (on 24H clock) at which the hourly FFMC drops below
+    the threshold of critical_ffmc.
+    Should only be called if critical_hour_start is not None.
+    """
+    if critical_hour_start is None:
+        return None
+    clock_time = critical_hour_start + 1.0    # increase time in increments of 1 hours
+    while get_afternoon_overnight_diurnal_ffmc(clock_time, solar_noon_ffmc) >= critical_ffmc:
+        clock_time += 1.0
+        if clock_time == 32:  # break if clock_time is now 08:00 of the next day
+            break
+    # subtract the hour that caused FFMC to drop below critical_ffmc
+    clock_time -= 1.0
+    if clock_time >= 24.0:
+        clock_time = clock_time - 24.0
+    return clock_time
+
+
+def get_critical_hours(  # pylint: disable=too-many-arguments
+        target_hfi: int, fuel_type: str, percentage_conifer: float,
+        percentage_dead_balsam_fir: float, bui: float,
+        grass_cure: float, crown_base_height: float,
+        solar_noon_ffmc: float, fmc: float, cfb: float, cfl: float,
+        wind_speed: float):
+    """ Determines the range of critical hours on a 24H clock.
+    Critical Hours describes the time range for the given day during which HFI will meet or exceed
+    hfi_target value. Critical hours are calculated by determining diurnally-adjusted FFMC values
+    that cause HFI >= target_hfi.
+    """
+    critical_ffmc, resulting_hfi = cffdrs.get_ffmc_for_target_hfi(
+        fuel_type, percentage_conifer, percentage_dead_balsam_fir, bui, wind_speed,
+        grass_cure, crown_base_height, solar_noon_ffmc, fmc, cfb, cfl, target_hfi)
+    logger.info('Critical FFMC %s, resulting HFI %s; target HFI %s', critical_ffmc,
+                resulting_hfi, target_hfi)
+    # Scenario 1: it's not possible for the HFI to reach target_hfi, in which case there will
```

Comment:
Also Scenario 2 has a `>=` check, suggesting it's possible for `resulting_hfi` to equal `target_hfi`?

Reply:
Or does each scenario describe the assumptions independently of other scenarios?

Reply:
Scenario 1 (`resulting_hfi < target_hfi`) - will happen when it's impossible to get a HFI value large enough to >= `target_hfi`, because FFMC influences the HFI value, and FFMC has an upper bound of 101. So basically, in this scenario the `resulting_hfi` would equal the resulting HFI when FFMC is set to 101.

Reply:
yes, each scenario is mutually exclusive of the others

Reply:
Can you make this the comment? I find it more clear: https://github.com/bcgov/wps/pull/1147#discussion_r673347923

Reply:
Scenario 2 has a `>=` check because I've reduced (drastically) the number of iterations required in `get_ffmc_for_target_hfi()` by allowing a margin of +/- 1%, rather than trying to get the _exact_ FFMC value at which HFI = `target_hfi`